stan/math/rev/core/profiling.hpp

```diff
@@ -0,0 +1,76 @@
+#ifndef STAN_MATH_REV_CORE_PROFILING_HPP
+#define STAN_MATH_REV_CORE_PROFILING_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/core/typedefs.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
+#include <stan/math/prim/err.hpp>
+#include <iostream>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+
+inline auto profile_start(std::string name, profile_map& profiles) {
+  profile_map::iterator it = profiles.find(name);
+  if (it == profiles.end()) {
+    profiles[name] = {};
+  }
+  if (profiles[name].meta.fwd_pass_active
```

Comment:
Save a reference to the underlying object and pass that around to avoid all the map lookups. There's always notes on the C++ docs about which operators invalidate references/iterators (https://en.cppreference.com/w/cpp/container/map/erase). I dunno if `iterators` are guaranteed to not need destructor called or I'd say copy those around.

Reply:
DOne.